### ISSUE TITLE
update-ssh-keys does not recognize options

### ISSUE THREAD

**Issue by mikiT** [coreos/bugs#2229]
# Issue Report #

### Container Linux Version ###

```
core@bad ~ $ cat /etc/os-release
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1576.1.0
VERSION_ID=1576.1.0
BUILD_ID=2017-10-26-0503
PRETTY_NAME="Container Linux by CoreOS 1576.1.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"

core@bad ~ $ ls -al /bin/update-ssh-keys 
-rwxr-xr-x. 1 root root 1211024 Oct 26 12:43 /bin/update-ssh-keys
...
BUG_REPORT_URL="https://issues.coreos.com"
```

### Environment ###

common x86 server

### Expected Behavior ###

```
core@good ~ $ cat /etc/os-release 
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1520.1.0
VERSION_ID=1520.1.0
BUILD_ID=2017-09-05-2146
PRETTY_NAME="Container Linux by CoreOS 1520.1.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"

core@good ~ $ ls -al /usr/bin/update-ssh-keys 
-rwxr-xr-x. 1 root root 4972 Aug 30 08:43 /usr/bin/update-ssh-keys
```

```
core@good ~ $ ls -al .ssh/authorized_keys.d/coreos-cloudinit
-rw-------. 1 core core 426 Nov  7 20:02 .ssh/authorized_keys.d/coreos-cloudinit
core@good ~ $ cat .ssh/authorized_keys.d/coreos-cloudinit 
from="192.168.1.0/24" ssh-rsa AAA...Qkn mikit@hoge

core@good ~ $ update-ssh-keys 
Updated /home/core/.ssh/authorized_keys
```

### Actual Behavior ###

```
core@bad ~ $ ls -al .ssh/authorized_keys.d/coreos-cloudinit 
-rw-r--r--. 1 core core 426 Nov  7 19:59 .ssh/authorized_keys.d/coreos-cloudinit
core@bad ~ $ cat .ssh/authorized_keys.d/coreos-cloudinit 
from="192.168.1.0/24" ssh-rsa AAA...Qkn mikit@hoge

core@bad ~ $ update-ssh-keys 
Error: failed to open authorized keys directory for user 'core'
Caused by: failed to parse public key
Caused by: invalid key format
Caused by: Invalid byte 45, offset 3.
```

### Reproduction Steps ###

We updated coreos from 1520.1.0 to 1576.1.0 last week.
After that, hoge now outputs an error.
We use `/bin/coreos-cloudinit` for configuration management.

This is extract from our `cloud-config.yml`.

```
(snip)
users:
  - name: core
    groups:
      - sudo
      - docker
    ssh_authorized_keys:
      - from="192.168.1.0/24" ssh-rsa AAA...Qkn mikit@hoge
(snip)
```

As this condition, we executed `/bin/coreos-cloudinit` then we saw error.

```
core@bad ~ $ sudo /bin/coreos-cloudinit -from-file=cloud-config.yml
(snip)
2017/11/07 20:38:42 Authorizing 32 SSH keys for user 'core'
2017/11/07 20:38:42 Failed to apply cloud-config: Call to update-ssh-keys failed with exit status 1:  Error: failed to open authorized keys directory for user 'core'
Caused by: failed to parse public key
Caused by: invalid key format
Caused by: Invalid byte 45, offset 3.
```

### Other Information ###

ref. sshd(8):
`authorized_keys` file format

**Comment by lucab** [comment]
Thanks for the report! It looks like the underlying parsing library is at fault [here](https://github.com/sdemos/openssh-keys/blob/5360d84e7d8e868c7a374ecf1a1bf6c7900a7605/src/lib.rs#L190) as it doesn't understand the first optional field. For reference, [format spec](https://manpages.debian.org/stretch/openssh-server/sshd.8.en.html#AUTHORIZED_KEYS_FILE_FORMAT) from manpage.

/cc @sdemos

**Comment by sdemos** [comment]
Yup, looks like the underlying ssh key parsing library didn't handle authorized_keys options at all. I added that support to the library, and added several tests to that effect. Thanks for finding this one. Sorry about the disruption!

I'm going to keep this open until it actually gets released in the os. Sorry about the noise.

**Comment by bgilbert** [comment]
This should be fixed in alpha 1591.0.0 and beta 1576.2.0, due soon.  Thanks for the report.
